File: tests/backend/unit/placeholder_rule_test.rb
# frozen_string_literal: true

require_relative '../../test_helper'
require 'services/rendering'

# TF-540 and TF-541 — what counts as a placeholder (AP-23, Requirements 8.2).
#
# **The rule exists twice**, here and in `frontend/src/util/rendering.js`, and
# it has to. The server substitutes when it renders and when it derives the
# variable set of a prompt; the browser does the same for the live preview and
# for the fields that appear while somebody types. Two notions of "what is a
# variable" would let a prompt carry metadata for something that never renders,
# or render something that has no metadata — and neither would raise anything.
#
# So both sides answer **one** table, `tests/fixtures/placeholder_cases.json`.
# Comparing the two patterns as text would not work: Ruby spells the rule
# `[[:alpha:]]` and JavaScript spells it `\p{L}`, and a comparison that had to
# tolerate that would end up comparing nothing.
class PlaceholderRuleTest < PromptAtelier::TestCase
  R = PromptAtelier::Rendering

  CASES = JSON.parse(
    File.read(File.expand_path('../../fixtures/placeholder_cases.json', __dir__), encoding: 'UTF-8')
  )['cases'].freeze

  def test_the_table_has_both_answers_in_it
    # Without this the case below would pass over an empty file, and the
    # comparison with the browser would be two sides agreeing about nothing.
    assert_operator CASES.size, :>, 10
    assert CASES.any? { |entry| entry['valid'] }
    assert CASES.any? { |entry| !entry['valid'] }
  end

  def test_tf541_accepts_exactly_what_requirements_8_2_accepts
    CASES.each do |entry|
      key = entry['key']
      result = R.render(body: "{{#{key}}}", variables: [{ key: key, value: 'X' }])

      assert_equal entry['valid'], result.text == 'X', "{{#{key}}}: #{entry['why']}"
    end
  end

  def test_whitespace_inside_the_braces_is_allowed
    assert_equal 'X', R.render(body: '{{ name }}', variables: [{ key: 'name', value: 'X' }]).text
  end

  # The keys a prompt derives its variables from have to be the same ones the
  # renderer substitutes — that is the other half of "one notion", and it lives
  # inside this file rather than between the two languages.
  def test_the_variable_set_is_derived_with_the_same_rule
    body = '{{año}} {{2fa}} {{prénom}}'

    assert_equal %w[año prénom], R.variable_keys(body)
  end

  # --- TF-540: what used to be silent ---------------------------------------

  # A placeholder the rule refuses is **reported**. Before AP-23 it was not
  # substituted, not reported as unknown and given no field to fill in — the
  # text simply kept it, and whoever wrote the prompt found out when they
  # pasted it into a model.
  def test_tf540_a_refused_placeholder_is_reported
    result = R.render(body: 'Hallo {{name}}, {{2fa}} und {{mi variable}}',
                      variables: [{ key: 'name', value: 'Martin' }])

    assert_equal ['2fa', 'mi variable'], result.rejected_keys
    assert_includes result.text, '{{2fa}}', 'and it stays in the text as it was written'
  end

  # Widening the key to Unicode letters does not remove the boundary, it moves
  # it — which is why the report is needed whatever the rule says.
  def test_tf540_a_name_in_another_language_is_a_variable_and_not_a_complaint
    result = R.render(body: 'Hola {{año}}, ciao {{città}}',
                      variables: [{ key: 'año', value: '2026' }, { key: 'città', value: 'Roma' }])

    assert_equal 'Hola 2026, ciao Roma', result.text
    assert_empty result.rejected_keys
    assert_empty result.unknown_keys
  end

  def test_tf540_an_escaped_placeholder_is_a_deliberate_literal
    result = R.render(body: 'Literal \\{{2fa}} bleibt')

    assert_empty result.rejected_keys
    assert_equal 'Literal {{2fa}} bleibt', result.text
  end

  # A stray `{{` must not pair with a `}}` far away and report the paragraph
  # between them as one enormous key.
  def test_tf540_braces_do_not_pair_across_lines
    assert_empty R.render(body: "offen {{ hier\nund dort }} zu").rejected_keys
  end
end

File: tests/frontend/placeholder_rule.test.js
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { render, rejectedKeys } from '../../frontend/src/util/rendering.js'

// TF-540 and TF-541 — what counts as a placeholder, and that both sides agree
// (AP-23, Requirements 8.2).
//
// **The rule exists twice.** The server substitutes when it renders and when
// it derives the variable set of a prompt; the browser does the same for the
// live preview and for the fields that appear while somebody types. Two
// notions of "what is a variable" would let a prompt carry metadata for
// something that never renders, or render something that has no metadata —
// and neither would raise anything.
//
// It had no test at all until now, which is how it survived being widened on
// one side only: `\p{L}` without the `u` flag is not a Unicode property, it is
// the four literal characters `p{L}`, so the pattern keeps matching — just
// something else. That is a difference nothing else in this project would see.

const HERE = path.dirname(fileURLToPath(import.meta.url))
const read = (...parts) => readFileSync(path.resolve(HERE, '..', '..', ...parts), 'utf8')

describe('The placeholder rule', () => {
  // TF-541. The case table lives in a file of its own, and the Ruby side
  // answers the very same one — that is what makes this a comparison rather
  // than two independent opinions.
  //
  // Compared through what the pattern **does**, not through its source: the
  // two languages spell a regular expression differently (`[[:alpha:]]` here,
  // `\p{L}` there), so a comparison of the text would either fail on the
  // spelling or be softened until it compared nothing.
  const { cases } = JSON.parse(read('tests', 'fixtures', 'placeholder_cases.json'))

  it('has a table to answer', () => {
    // Without this the loop below would pass over an empty file.
    expect(cases.length).toBeGreaterThan(10)
    expect(cases.some(({ valid }) => valid)).toBe(true)
    expect(cases.some(({ valid }) => !valid)).toBe(true)
  })

  it('accepts exactly what Requirements 8.2 accepts', () => {
    for (const { key, valid, why } of cases) {
      const substituted = render({ body: `{{${key}}}`, variables: [{ key, value: 'X' }] }).text

      expect(substituted === 'X', `{{${key}}}: ${why}`).toBe(valid)
    }
  })

  it('allows whitespace inside the braces', () => {
    expect(render({ body: '{{ name }}', variables: [{ key: 'name', value: 'X' }] }).text).toBe('X')
  })

  // TF-540. The half that used to be silent: a placeholder the rule refuses is
  // **reported**. Before this, `{{2fa}}` was not substituted, not reported as
  // unknown and given no field — the text simply kept it, and whoever wrote
  // the prompt found out when they pasted it into a model.
  it('reports what it refuses instead of leaving it standing', () => {
    const rejected = rejectedKeys('{{name}} {{2fa}} und {{mi variable}} und {{año}}')

    expect(rejected).toEqual(['2fa', 'mi variable'])
  })

  it('takes an escaped placeholder for a deliberate literal, not a mistake', () => {
    expect(rejectedKeys(String.raw`Literal \{{2fa}} bleibt`)).toEqual([])
  })

  // The counter-check: a valid one is never reported, or the warning would
  // stand under every prompt that has variables at all.
  it('reports nothing for a text whose placeholders are all valid', () => {
    expect(rejectedKeys('{{año}} {{prénom}} {{a_1}} {{ name }}')).toEqual([])
  })

  // A stray `{{` must not pair with a `}}` far away and report the paragraph
  // between them as one enormous key.
  it('does not pair braces across lines', () => {
    expect(rejectedKeys('offen {{ hier\nund dort }} zu')).toEqual([])
  })
})
